### ISSUE TITLE
Git Credential Core MFA question

### ISSUE THREAD

**Issue by toniBisson** [GitCredentialManager/git-credential-manager#303]
Just configuring the Git Credential core on my work laptop... has to go through a proxy.   Two Questions

1.  Are we able to use MFA without generating a PAT?  
2. Documentation says a PAT would be generated on our behalf but I was actually prompted for the Pat. So I manually created one and entered the passcode.
I am able to login now.  I would really like to just authenticate thru MFA without the PAT.  Let me know if it is possible.

**Comment by kyle-rader** [comment]
Hi @toniBisson, 

> Are we able to use MFA without generating a PAT?

Functionality of regularly using AAD MFA for Azure Devops is underway in #294 - so not available just yet. 

> PAT would be generated on our behalf but I was actually prompted for the Pat 

On first use that it expected, as GCM-Core still has to do an AAD (MFA is needed) authentication step in order to generate a PAT to store and use from then on (until it needs to regen a PAT). So this behavior is expected.

**Comment by oarm14** [comment]
@kyle-rader is a web-like form supposed to appear to enter the credentials on the first time? I get an open ssh window.  I'm in the same boat as Toni as far as proxy and tried the same things.

![image](https://user-images.githubusercontent.com/59032199/111804624-a179e100-88a6-11eb-8591-4ec1df61dcfb.png)

**Comment by PaulCormier** [comment]
I work with Toni and can add some details to what's happening. We are using an HTTP proxy with basic authentication.

I turned on GCM_TRACE and found the following relevant (IMHO) lines:

> 12:16:32.931564 ...pClientFactory.cs:60 trace: [CreateClient] HTTP client is using the configured proxy.
12:16:57.514033 ...sHostProvider.cs:208 trace: [GeneratePersonalAccessTokenAsync] Acquired Azure access token. Account='{Redacted}' Token='********'
12:16:57.514033 ...sHostProvider.cs:217 trace: [GeneratePersonalAccessTokenAsync] Creating Azure DevOps PAT with scopes 'vso.code_write, vso.packaging'...
12:16:57.516059 ...DevOpsRestApi.cs:119 trace: [CreatePersonalAccessTokenAsync] Getting Azure DevOps Identity Service endpoint...
12:16:57.519052 ...DevOpsRestApi.cs:169 trace: [GetIdentityServiceUriAsync] HTTP: GET https://dev.azure.com/{Redacted}/_apis/ServiceDefinitions/LocationService2/{Redacted}?api-version=1.0
12:16:57.735442 ...DevOpsRestApi.cs:173 trace: [GetIdentityServiceUriAsync] HTTP: Response 200 [OK]
12:16:57.737437 ...DevOpsRestApi.cs:121 trace: [CreatePersonalAccessTokenAsync] Identity Service endpoint is 'https://spsprodcca1.vssps.visualstudio.com/{Redacted}/'.
12:16:57.737437 ...DevOpsRestApi.cs:125 trace: [CreatePersonalAccessTokenAsync] HTTP: POST https://spsprodcca1.vssps.visualstudio.com/{Redacted}/_apis/token/sessiontokens?api-version=1.0&tokentype=compact

On the console I'm seeing:
> fatal: An error occurred while sending the request.
fatal: The remote server returned an error: (407) Proxy Authentication Required.
fatal: Failed to write item to store. [0x6f7]
fatal: The stub received bad data

I used Fiddler to capture the requests, and I see the first 407 without the Proxy-Authorization header, followed by another request with the header, but it only includes my username, and no password.

I'm using the latest GCMCore, running on Windows 10, and other functionality is working correctly with the proxy.

Let me know if you need more information, or to have me test something.

**Comment by PaulCormier** [comment]
Further testing with `GCM_TRACE_SECRETS=1` has revealed that the password is not being passed to `TryCreateProxy`. When I include my HTTP proxy password as part of the proxy URL, it is passed in, but if I only enter my user id, it is blank. I am prompted for a password though when I run a git command, but I can see it's already logged no password by that point.

> 15:17:43.380326 ...pClientFactory.cs:54 trace: [CreateClient] Creating new HTTP client instance...
15:17:43.569821 ...ClientFactory.cs:151 trace: [TryCreateProxy] Created a WebProxy instance:
15:17:43.569821 ...ClientFactory.cs:152 trace: [TryCreateProxy] 	address=http://http-proxy:8080/
15:17:43.569821 ...ClientFactory.cs:152 trace: [TryCreateProxy] 	username={Redacted}
15:17:43.569821 ...ClientFactory.cs:152 trace: [TryCreateProxy] 	password=
15:17:43.569821 ...pClientFactory.cs:60 trace: [CreateClient] HTTP client is using the configured proxy.

**Comment by mjcheetham** [comment]
Hi @PaulCormier and @toniBisson!

> I'm using the latest GCMCore, running on Windows 10, and other functionality is working correctly with the proxy.

Can I ask which GCM Core version this is specifically?

I've just set-up a local proxy (mitmproxy) with GCM Core 2.0.452, and when tracing I see both the proxy username (`alice`) and password (`letmein123`) are being used:

```shell
$ GCM_TRACE=1 GCM_TRACE_SECRETS=1 GIT_SSL_NO_VERIFY=1 git fetch
09:23:39.328749 ...pplicationBase.cs:76 trace: [RunAsync] Tracing of secrets is enabled. Trace output may contain sensitive information.
09:23:39.351173 ...er/Application.cs:80 trace: [RunInternalAsync] Version: 2.0.452.3248
09:23:39.351220 ...er/Application.cs:81 trace: [RunInternalAsync] Runtime: .NET 5.0.6
09:23:39.351232 ...er/Application.cs:82 trace: [RunInternalAsync] Platform: macOS (x86-64)
09:23:39.351289 ...er/Application.cs:83 trace: [RunInternalAsync] AppPath: /usr/local/share/gcm-core/git-credential-manager-core
09:23:39.351400 ...er/Application.cs:84 trace: [RunInternalAsync] Arguments: get
09:23:39.410206 ...GitCommandBase.cs:35 trace: [ExecuteAsync] Start 'get' command...
09:23:39.417786 ...GitCommandBase.cs:49 trace: [ExecuteAsync] Detecting host provider for input:
09:23:39.419005 ...GitCommandBase.cs:50 trace: [ExecuteAsync] 	protocol=https
09:23:39.419038 ...GitCommandBase.cs:50 trace: [ExecuteAsync] 	host={org}.visualstudio.com
09:23:39.419046 ...GitCommandBase.cs:50 trace: [ExecuteAsync] 	username={user}
09:23:39.459629 ...viderRegistry.cs:149 trace: [GetProviderAsync] Performing auto-detection of host provider.
09:23:39.461672 ...viderRegistry.cs:162 trace: [GetProviderAsync] Checking against 3 host providers registered with priority 'Normal'.
09:23:39.462535 ...GitCommandBase.cs:52 trace: [ExecuteAsync] Host provider 'Azure Repos' was selected.
09:23:39.483382 ...osHostProvider.cs:87 trace: [GetCredentialAsync] Looking for existing credential in store with service=https://{org}.visualstudio.com account={user}...
09:23:39.520488 ...osHostProvider.cs:92 trace: [GetCredentialAsync] No existing credentials found.
09:23:39.520535 ...osHostProvider.cs:95 trace: [GetCredentialAsync] Creating new credential...
09:23:39.522330 ...sHostProvider.cs:196 trace: [GeneratePersonalAccessTokenAsync] Determining Microsoft Authentication Authority...
09:23:39.544978 ...eDevOpsRestApi.cs:45 trace: [GetAuthorityAsync] HTTP: HEAD https://{org}.visualstudio.com/
09:23:39.546634 ...pClientFactory.cs:54 trace: [CreateClient] Creating new HTTP client instance...
09:23:39.584036 ...ClientFactory.cs:115 trace: [TryCreateProxy] Using a deprecated proxy configuration.
warning: Using a deprecated proxy configuration. See https://aka.ms/gcmcore-httpproxy for more information.
09:23:39.584630 ...ClientFactory.cs:151 trace: [TryCreateProxy] Created a WebProxy instance:
09:23:39.584733 ...ClientFactory.cs:152 trace: [TryCreateProxy] 	address=http://localhost:8080/
09:23:39.584769 ...ClientFactory.cs:152 trace: [TryCreateProxy] 	username=alice
09:23:39.584781 ...ClientFactory.cs:152 trace: [TryCreateProxy] 	password=letmein123
09:23:39.584789 ...pClientFactory.cs:60 trace: [CreateClient] HTTP client is using the configured proxy.
```

Can I ask which configuration options you are using to configure the proxy?

- Git configuration:
   - `http.Proxy` or `https.proxy`
   - `credential.httpProxy` or `credential.httpsProxy` (deprecated)
- Environment variables:
   - `HTTP_PROXY`, `HTTPS_PROXY`, `ALL_PROXY`
   - `GCM_HTTP_PROXY` (deprecated)

Does the password contain any "special" characters? That is, any characters that should be URL encoded like `@` to `%40`, `:` to `%3A`, or ` ` (space) to `%20`?

**Comment by PaulCormier** [comment]
I was using version 2.0.394.50751, and configuring the proxy via the http.proxy config parameter. The password does contain a "special" character, but as it is being supplied into a password prompt (not in the config file), and is base64 encoded, I don't think it's a significant factor.

I've upgraded GCMCore to 2.0.452.3248, and tried again. Unfortunately, we've changed our corporate proxy to not require authentication on the host: spsprodcca1.vssps.visualstudio.com. This makes reproducing the problem difficult. I believe I was able to reproduce the issue by using Fiddler to return a 407 on the first request to spsprodcca1.vssps.visualstudio.com. GCMCore responds with a Proxy-Authorization header which only has the user name, no password. In the trace log, I see it pulling up the correct credentials, but then the call to TryCreateProxy doesn't seem to use them. This is the same behaviour I was seeing before, and if the proxy actually required credentials for that host, the missing password would be a problem.

> 10:16:57.868212 ...pplicationBase.cs:76 trace: [RunAsync] Tracing of secrets is enabled. Trace output may contain sensitive information.
10:16:57.882176 ...\Application.cs:80   trace: [RunInternalAsync] Version: 2.0.452.3248
10:16:57.882176 ...\Application.cs:81   trace: [RunInternalAsync] Runtime: .NET Framework 4.0.30319.42000
10:16:57.882176 ...\Application.cs:82   trace: [RunInternalAsync] Platform: Windows (x86-64)
10:16:57.882176 ...\Application.cs:83   trace: [RunInternalAsync] AppPath: git-credential-manager-core
10:16:57.882176 ...\Application.cs:84   trace: [RunInternalAsync] Arguments: store
10:16:57.935996 ...GitCommandBase.cs:35 trace: [ExecuteAsync] Start 'store' command...
10:16:57.944972 ...GitCommandBase.cs:49 trace: [ExecuteAsync] Detecting host provider for input:
10:16:57.945970 ...GitCommandBase.cs:50 trace: [ExecuteAsync] 	protocol=http
10:16:57.945970 ...GitCommandBase.cs:50 trace: [ExecuteAsync] 	host=localhost:8888
10:16:57.945970 ...GitCommandBase.cs:50 trace: [ExecuteAsync] 	username={redacted username}
10:16:57.945970 ...GitCommandBase.cs:50 trace: [ExecuteAsync] 	password={redacted password}
10:16:58.072632 ...viderRegistry.cs:149 trace: [GetProviderAsync] Performing auto-detection of host provider.
10:16:58.073660 ...viderRegistry.cs:162 trace: [GetProviderAsync] Checking against 3 host providers registered with priority 'Normal'.
10:16:58.075655 ...viderRegistry.cs:154 trace: [GetProviderAsync] Querying remote URL for host provider auto-detection.
10:16:58.075655 ...pClientFactory.cs:54 trace: [CreateClient] Creating new HTTP client instance...
10:16:58.197300 ...ClientFactory.cs:151 trace: [TryCreateProxy] Created a WebProxy instance:
10:16:58.197300 ...ClientFactory.cs:152 trace: [TryCreateProxy] 	address=http://localhost:8888/
10:16:58.197300 ...ClientFactory.cs:152 trace: [TryCreateProxy] 	username={redacted username}
10:16:58.197300 ...ClientFactory.cs:152 trace: [TryCreateProxy] 	password=
10:16:58.197300 ...pClientFactory.cs:60 trace: [CreateClient] HTTP client is using the configured proxy.

Perhaps there's an issue when the proxy only requires authentication for the spsprodcca1.vssps.visualstudio.com host? We had previously disabled proxy authentication on a few other hosts pertaining to Azure. If GCMCore didn't get challenged for a password early on in the process, maybe it's not getting initialized correctly for later on?
